### PR TITLE
GAEv1 support.

### PR DESCRIPTION
**Pull request by benwulfe**
An external ThreadFactory requires us to refactor WebSocket and WebSocketWriter so that they do not subclass Thread.

### CODE REVIEW

**Comment by benwulfe** [comment]
@jdimond This is pretty similar to the change in the firebase client.  It allows an external ThreadFactory to be used.  It's a bit more complicated only because two classes derive from Thread, which requires a bit of code shuffling, but it actually isn't that bad.

**Comment by jdimond** [comment]
@benwulfe thanks for this pull request. I'm sure people are going to be happy about this. We maybe want to consider actually having a GaePlatform class that takes care of all of this and also shows up differently in our stats etc. If we can detect GAE and solve the ThreadFactory problem for our devs it will be a lot easier to use...

**Comment by benwulfe** [comment]
thanks for taking time to review Jonny.
re: platform yeah, that was my first instinct and how I did my current blog sample.  The issue is that you'd need to define the platform outside of the base jvm client, which means more code for the user and more surface area for them to muck with firebase internals.
That's why I scoped it down to only sending in a ThreadFactory.

File: src/main/java/com/firebase/tubesock/WebSocket.java [hunk 1]
```diff
@@ -52,6 +54,31 @@
     private final WebSocketHandshake handshake;
     private final int clientId = clientCount.incrementAndGet();
 
+    private Thread innerThread = null;
+    private static ThreadFactory threadFactory = Executors.defaultThreadFactory();
+    private static ThreadInitializer intializer;
+
+    public static ThreadFactory getThreadFactory() {
```
Comment (jdimond) [inline src/main/java/com/firebase/tubesock/WebSocket.java]: should probably be package private

File: src/main/java/com/firebase/tubesock/WebSocket.java [hunk 2]
```diff
@@ -52,6 +54,31 @@
...
 
+    private Thread innerThread = null;
+    private static ThreadFactory threadFactory = Executors.defaultThreadFactory();
+    private static ThreadInitializer intializer;
+
+    public static ThreadFactory getThreadFactory() {
+        return threadFactory;
+    }
+
+    public static ThreadInitializer getIntializer() {
```
Comment (jdimond) [inline src/main/java/com/firebase/tubesock/WebSocket.java]: should probably be package private

File: src/main/java/com/firebase/tubesock/WebSocket.java [hunk 3]
```diff
@@ -344,9 +292,97 @@ private void verifyHost(SSLSocket socket, String host) throws SSLException {
      */
     public void blockClose() throws InterruptedException {
         // If the thread is new, it will never run, since we closed the connection before we actually connected
-        if (writer.getState() != Thread.State.NEW) {
-            writer.join();
+        if (writer.getInnerThread().getState() != Thread.State.NEW) {
```
Comment (jdimond) [inline src/main/java/com/firebase/tubesock/WebSocket.java]: there is a subtle race condition here, that if you call blockClose() immediately after start() you might call getInnerThread() on two different threads resulting in two different Threads created. I would propose the following changes:
a) Make a runWriter and runReader method respectively
b) Create the threads in the constructors and assign them to a local final variable
c) Make getInnerThread package private
d) Return a local final variable in the getInnerThread

**Comment by benwulfe** [comment]
@jonny great catch, thanks.  I've updated the PR.

**Pull request merged**